src/Credentials/AssumeRoleCredentialProvider.php

```diff
@@ -0,0 +1,82 @@
+<?php
+namespace Aws\Credentials;
+
+use Aws\Exception\AwsException;
+use Aws\Exception\CredentialsException;
+use Aws\Result;
+use Aws\Sts\StsClient;
+use GuzzleHttp\Promise;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Psr7\Response;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Credential provider that provides credentials via assuming a role
+ * More Information, see: http://docs.aws.amazon.com/aws-sdk-php/v3/api/api-sts-2011-06-15.html#assumerole
+ */
+class AssumeRoleCredentialProvider
+{
+    const ERROR_MSG = "Missing required 'AssumeRoleCredentialProvider' configuration option: ";
+
+    /** @var callable */
+    private $client;
+
+    /** @var array */
+    private $assumeRoleParams;
+
+    /**
+     * The constructor requires following configure parameters:
+     *  - client: a customized Sts Client
+     *
+     *  paramters used to construct a default StsClient(which will use the latest version):
+     *    - region: the StsClient region for assume role
+     *    - credentials: a CredentialProvider used to retrieve assume role
+     *
+     *  - assume_role_params: Parameters used to make assumeRole call
+     *
+     * @param array $config Configuration options
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(array $config = [])
+    {
+        if (!isset($config['region'])) {
+            throw new \InvalidArgumentException(self::ERROR_MSG . "'region'.");
+        }
+        if (!isset($config['credentials'])) {
+            throw new \InvalidArgumentException(self::ERROR_MSG . "'credentials'.");
+        }
+        if (!isset($config['assume_role_params'])) {
+            throw new \InvalidArgumentException(self::ERROR_MSG . "'assume_role_params'.");
+        }
+
+        $defaultClientArgs = [
+            'region' => $config['region'],
+            'version' => 'latest',
+            'credentials' => $config['credentials'],
+        ];
+        $this->client = isset($config['client'])
+            ? $config['client']
+            : new StsClient($defaultClientArgs);
+        $this->assumeRoleParams = $config['assume_role_params'];
+    }
+
+    /**
+     * Loads assume role credentials.
+     *
+     * @return PromiseInterface
+     */
+    public function __invoke()
+    {
+        $client = $this->client;
+        return Promise\promise_for($client->assumeRoleAsync($this->assumeRoleParams))
```

Comment:
You don't need a promise_for when you know the return value of `assumeRoleAsync` will always be a promise.